tty/gr716-uart/gr716-uart.c

```diff
@@ -0,0 +1,463 @@
+					msg.o.attr.err = EOK;
+					break;
+				}
+
+				msg.o.attr.err = -EINVAL;
+				break;
+
+			case mtDevCtl:
+				uart_ioctl(port, &msg);
+				break;
+		}
+
+		msgRespond(port, &msg, rid);
+	}
+}
+
+
+static void uart_klogClbk(const char *data, size_t size)
+{
+	libtty_write(&uart_common.uart.tty, data, size, 0);
+}
+
+
+static void uart_mkDev(unsigned int id)
+{
+	char path[12];
+
+	snprintf(path, sizeof(path), "/dev/uart%u", id);
+	if (create_dev(&uart_common.uart.oid, path) < 0) {
+		debug("gr716-uart: Cannot create device file.\n");
+		return;
+	}
+
+	if (id == UART_CONSOLE_USER) {
+		libklog_init(uart_klogClbk);
+
+		if (create_dev(&uart_common.uart.oid, _PATH_CONSOLE) < 0) {
+			debug("gr716-uart: Cannot create device file.\n");
+		}
+	}
+}
+
+
+static int uart_setPin(uint8_t pin)
+{
+	platformctl_t ctl;
+
+	ctl.action = pctl_set;
+	ctl.type = pctl_iocfg;
+
+	ctl.iocfg.pin = pin;
+	ctl.iocfg.pulldn = 0;
+	ctl.iocfg.pullup = 0;
+	ctl.iocfg.opt = 0x1;
+
+	switch (pin) {
+		case UART0_TX:
+		case UART1_TX:
+			/* pins used for SRAM */
+			return -EINVAL;
+		case UART2_TX:
+		case UART3_TX:
+		case UART4_TX:
+		case UART5_TX:
+			ctl.iocfg.dir = GPIO_DIR_OUT;
+			break;
+		case UART0_RX:
+		case UART1_RX:
+			/* pins used for SRAM */
+			return -EINVAL;
+		case UART2_RX:
+		case UART3_RX:
+		case UART4_RX:
+		case UART5_RX:
+			ctl.iocfg.dir = GPIO_DIR_IN;
+			break;
+		default:
+			return -EINVAL;
+	}
+
+	return platformctl(&ctl);
+}
+
+
+static int uart_cguInit(unsigned int cgudev)
+{
+	platformctl_t ctl;
+
+	ctl.action = pctl_set;
+	ctl.type = pctl_cguctrl;
+
+	ctl.cguctrl.state = enable;
+	ctl.cguctrl.cgu = cgu_primary;
+	ctl.cguctrl.cgudev = cgudev;
+
+	return platformctl(&ctl);
+}
+
+
+static int uart_init(unsigned int n, speed_t baud, int raw)
+{
+	libtty_callbacks_t callbacks;
+	uart_t *uart = &uart_common.uart;
+
+	if (uart_setPin(info[n].rxPin) < 0 || uart_setPin(info[n].txPin) < 0 || uart_cguInit(info[n].cgudev) < 0) {
+		return -EINVAL;
+	}
+
+	uart->base = info[n].base;
+
+	callbacks.arg = uart;
+	callbacks.set_cflag = uart_setCFlag;
+	callbacks.set_baudrate = uart_setBaudrate;
+	callbacks.signal_txready = uart_signalTXReady;
+
+	if (libtty_init(&uart->tty, &callbacks, _PAGE_SIZE, baud) < 0) {
+		return -ENOENT;
+	}
+
+	if (condCreate(&uart->cond) != EOK) {
+		libtty_close(&uart->tty);
+		libtty_destroy(&uart->tty);
+		return -ENOENT;
+	}
+
+	if (mutexCreate(&uart->lock) != EOK) {
+		libtty_close(&uart->tty);
+		libtty_destroy(&uart->tty);
+		resourceDestroy(uart->cond);
+		return -ENOENT;
+	}
+
+	/* Set raw mode */
+	if (raw == 1) {
+		libtty_set_mode_raw(&uart->tty);
+	}
+
+	/* normal mode, 1 stop bit, no parity, 8 bits */
+	uart_setCFlag(uart, &uart->tty.term.c_cflag);
+
+	uart->tty.term.c_ispeed = uart->tty.term.c_ospeed = baud;
+	uart_setBaudrate(uart, baud);
+	*(uart->base + uart_ctrl) = RX_INT | RX_EN | TX_EN;
+
+	beginthread(uart_intThread, 4, &uart->stack, sizeof(uart->stack), (void *)uart);
```

Comment:
4 is a very low priority for interrupt thread
```suggestion
	beginthread(uart_intThread, 2, &uart->stack, sizeof(uart->stack), (void *)uart);
```